README.md
## date-type [![npm][npm-svg]][npm]

[![Build Status][travis-svg]][travis]
[![Code Climate][codeclimate-svg]][codeclimate]
[![Test Coverage][codeclimate-test-svg]][codeclimate-test]
[![downloads][npm-download-svg]][npm]
[![license][npm-license-svg]][npm]

[npm]: https://npmjs.org/package/date-type
[npm-svg]: https://img.shields.io/npm/v/date-type.svg
[npm-download-svg]: https://img.shields.io/npm/dm/date-type.svg
[npm-license-svg]: https://img.shields.io/npm/l/date-type.svg
[travis-svg]: https://img.shields.io/travis/snowyu/date-type.js/master.svg
[travis]: http://travis-ci.org/snowyu/date-type.js
[codeclimate-svg]: https://codeclimate.com/github/snowyu/date-type.js/badges/gpa.svg
[codeclimate]: https://codeclimate.com/github/snowyu/date-type.js
[codeclimate-test-svg]: https://codeclimate.com/github/snowyu/date-type.js/badges/coverage.svg
[codeclimate-test]: https://codeclimate.com/github/snowyu/date-type.js/coverage

The date type.

## Usage

```js
//register the date type to the TypeInfo..
import { Type, DateType } from 'date-type'

expect(Type.get('Date')).toStrictEqual(DateType)

var today = DateType.today()
//var today = new DateType(new Date()).clearTime() //it's the same.
//var today = DateType.now().clearTime() //it's the same.

var nextDay = today.clone().addDays(1)
var FutureType = new Type('date', {min: nextDay})

console.log(today)
//<type "Date": "value":"2016-02-28T04:13:46.352Z">
console.log(today.isValid())
//=true
console.log(today.toJson())
//='"2016-02-28T04:13:46.352Z"'
console.log(today.toJson({withType:true}))
//='{"value":"2016-02-28T04:13:46.352Z","name":"Date"}'
console.log(FutureType.isValid(nextDay))
//=true
console.log(FutureType.isValid(today))
//=true
var n = nextDay.clone();
n.assign('aaa')
//=TypeError: "aaa" is an invalid Date
console.log(nextDay > today)
//=true
```

## API

* The DateType Class
  * Methods:
    * `now()`: Gets a date type object that is set to the current date time.
    * `today()`: Gets a date type object that is set to the current date.
      * The time is set to the start of the day (00:00 or 12:00 AM).
    * `clearTime(date)`: Resets the time of the Date object to 12:00 AM (00:00), which is the start of the day.
    * `isLeapYear(aYear)`: Determines if the year whether a LeapYear.
    * `getDaysInMonth(aYear, aMonth)`: Gets the number of days in the month, given a year and month value. Automatically corrects for LeapYear.
* The DateType Value Class
  * Methods:
    * `addYears(value)`: Adds the specified number of years to this instance.
    * `addMonths(value)`: Adds the specified number of months to this instance.
    * `addWeeks(value)`: Adds the specified number of weeks to this instance.
    * `addDays(value)`: Adds the specified number of days to this instance.
    * `clearTime()`: Resets the time of this instance to 12:00 AM (00:00), which is the start of the day.
    * `isLeapYear()`: Determines this date whether a LeapYear.
    * `getDaysInMonth()`: Gets the number of days in the month. Automatically corrects for LeapYear.
    * `equals(value)`: Compares this instance to another Date object and returns true if they are equal.
    * `compareTo(value)`: Compares this instance to a Date object and returns an number indication of their relative values.
      * `return`: -1 = this is lessthan date. 0 = values are equal. 1 = this is greaterthan date.

See [abstract-type](https://github.com/snowyu/abstract-type.js)

## TODO


## License

MIT


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.0.0-alpha.1](https://github.com/snowyu/date-type.js/compare/v2.0.0-alpha.0...v2.0.0-alpha.1) (2021-06-07)


### Bug Fixes

* missing number-type package ([6400188](https://github.com/snowyu/date-type.js/commit/6400188aa432cb98fc1276a5e41db7417d0bd6c6))
* optimal toValue method ([07c03b5](https://github.com/snowyu/date-type.js/commit/07c03b590fef06fc506ac8b01aa149a8b0198ca1))

## [2.0.0-alpha.0](https://github.com/snowyu/date-type.js/compare/v0.1.2...v2.0.0-alpha.0) (2021-05-10)


src/date-type.spec.ts
import 'jest-extended'
// import isFunction from 'util-ex/lib/is/type/function'

import { DateType, Type } from './index'

const minDate = 1264953600000 // 2010/1/1
const lastDate = 1456611449302
const todayDate = Date.now()

describe('DateType', () => {
  const dt = new DateType()
  it('should try to get DateType type', () => {
    const v = new Type(new Date())
    expect(v).toBeInstanceOf(DateType)
  })

  describe('Static Members', () => {
    it('should be exist Date type', () => {
      expect(dt).toBeInstanceOf(DateType)
      expect(DateType.pathArray()).toEqual(['type', 'AbstractNumber', dt.name])
    })

    describe('range', () => {
      it('should not set the wrong range of Date', () => {
        expect(() => new DateType({ max: 'as' })).toThrow(
          'the max should be a ' + dt.name
        )
        expect(() => new DateType(null, { max: 'as' })).toThrow(
          'the max should be a ' + dt.name
        )

        expect(() => new DateType(null, { min: 'as' })).toThrow(
          'the min should be a ' + dt.name
        )
        expect(() => new DateType({ min: 'as' })).toThrow(
          'the min should be a ' + dt.name
        )

        expect(
          () => new DateType(null, { min: todayDate, max: lastDate })
        ).toThrow('the max should be greater than min')
        expect(() => new DateType({ min: todayDate, max: lastDate })).toThrow(
          'the max should be greater than min'
        )

        const result = dt.clone()
        result.max = lastDate
        expect(() => (result.min = todayDate)).toThrow(
          'the min should be less than max'
        )
      })
      it('should limit the range of DateType value', () => {
        const n = new DateType({ min: 2, max: 6 })
        expect(n.min.valueOf()).toEqual(2)
        expect(n.max.valueOf()).toEqual(6)
        expect(n.validate(4)).toBeTrue()
        expect(n.validate(2)).toBeTrue()
        expect(n.validate(6)).toBeTrue()
        expect(n.validate.bind(n, 1)).toThrow('an invalid ' + dt.name)
        expect(n.validate.bind(n, 7)).toThrow('an invalid ' + dt.name)
        expect(n.isValid(0)).toBeFalse()
        expect(n.isValid(7)).toBeFalse()
      })
      it('should limit max DateType value', () => {
        const n = new DateType(null, { max: 6 })
        expect(n.validate(1)).toBeTrue()
        expect(n.validate(4)).toBeTrue()
        expect(n.validate(2)).toBeTrue()
        expect(n.validate(6)).toBeTrue()
        expect(n.validate.bind(n, 7)).toThrow('an invalid ' + dt.name)
        expect(n.isValid(0)).toBeTrue()
        expect(n.isValid(7)).toBeFalse()
        expect(n.isValid(8)).toBeFalse()
      })
      it('should limit min DateType value', () => {
        const n = new DateType({ min: 2 })
        expect(n.validate(4)).toBeTrue()
        expect(n.validate(2)).toBeTrue()
        expect(n.validate(6)).toBeTrue()
        expect(n.validate(7)).toBeTrue()
        expect(n.validate.bind(n, 1)).toThrow('an invalid ' + dt.name)
        expect(n.isValid(0)).toBeFalse()
        expect(n.isValid(-1)).toBeFalse()
      })
    })

    describe('.validate', () => {
      it('should validate string DateType value', () => {
        expect(dt.validate(123)).toBeTrue()
        expect(dt.validate('2016/2/28')).toBeTrue()
        expect(dt.validate(new Date())).toBeTrue()
        expect(dt.validate.bind(dt, 'dsd')).toThrow('an invalid ' + dt.name)
      })
    })

    describe('.clone', () => {
      it('should clone type', () => {
        const t = new DateType({ min: minDate, max: Date.now() + 1000 })
        const result = t.clone()
        expect(t.isSame(result)).toBeTrue()
      })
    })
    describe('.now', () => {
      it('should return a current datetime value object', () => {
        const first = new Date()
        const t = DateType.now()
        expect(t).toBeInstanceOf(DateType)
        expect(t.valueOf()).toBeGreaterThanOrEqual(first.valueOf())
        expect(t.valueOf()).toBeLessThanOrEqual(new Date().valueOf())
      })
    })

    describe('.today', () => {
      it('should return a current today value object', () => {
        const today = new Date()
        const t = DateType.today()
        expect(t).toBeInstanceOf(DateType)
        expect(t.valueOf()).toBeLessThanOrEqual(today.valueOf())
        expect(t.getHours()).toEqual(0)
        expect(t.getMinutes()).toEqual(0)
        expect(t.getSeconds()).toEqual(0)
        expect(t.getMilliseconds()).toEqual(0)
        DateType.clearTime(today)
        expect(t.valueOf()).toStrictEqual(today.valueOf())
      })
    })
    describe('.isLeapYear', () => {
      it('should check a year whether be a leap', () => {
        expect(DateType.isLeapYear(2012)).toBeTrue()
        expect(DateType.isLeapYear(2016)).toBeTrue()
        expect(DateType.isLeapYear(2020)).toBeTrue()
        expect(DateType.isLeapYear(2024)).toBeTrue()
        expect(DateType.isLeapYear(3008)).toBeTrue()
        for (const i of [2013, 2014, 2015]) {
          expect(DateType.isLeapYear(i)).toBeFalse()
        }
        for (const i of [2017, 2018, 2019]) {
          expect(DateType.isLeapYear(i)).toBeFalse()
        }
        for (const i of [2021, 2022, 2023]) {
          expect(DateType.isLeapYear(i)).toBeFalse()
        }
      })
    })
    describe('.getDaysInMonth', () => {
      it('should get the days in a specified month', () => {
        expect(DateType.getDaysInMonth(2016, 0)).toEqual(31)
        expect(DateType.getDaysInMonth(2016, 1)).toEqual(29)
      })
    })
    describe('.equals', () => {
      it('should Compares the two Date whether be equal.', () => {
        // expect(DateType.equals(2, 1)).toBeFalse()
        // expect(DateType.equals(1, 1)).toBeTrue()
        expect(DateType.equals(new DateType(1), 1)).toBeTrue()
        // expect(DateType.equals(new DateType(1), new DateType(1))).toBeTrue()
        // expect(DateType.equals.bind(DateType, 'abc', 's')).toThrow('abc - s')
      })
    })
    describe('.compare', () => {
      it('should Compares the two Date.', () => {
        expect(DateType.compare(2, 1)).toEqual(1)
        expect(DateType.compare(1, 1)).toEqual(0)
        expect(DateType.compare(new DateType(1), 2)).toEqual(-1)
        expect(DateType.compare(new DateType(2), new DateType(1))).toEqual(1)
        expect(DateType.compare.bind(DateType, 'abc', 's')).toThrow('abc - s')
      })
    })
    describe('.toISOString', () => {
      it('should toISOString', () => {
        const dt = new Date()
        expect(DateType.toISOString(dt)).toStrictEqual(dt.toISOString())
        expect(DateType.toISOString(new DateType(dt))).toStrictEqual(
          dt.toISOString()
        )
        expect(DateType.toISOString('2020-10-10')).toStrictEqual(
          '2020-10-10T00:00:00.000Z'
        )
      })
    })
    describe('.toValue', () => {
      it('should convert value to Date', () => {
        const now = Date.now()
        const dt = DateType.toValue(DateType.now())
        expect(dt).toBeNumber()
        expect(dt.valueOf()).toBeWithin(now, now + 10)
      })
    })
  })

  describe('Instance Members', () => {
    describe('clone', () => {
      it('should clone value', () => {
        const t = DateType.now()
        const result = t.clone()
        expect(t.isSame(result)).toBeTrue()
      })
    })

    describe('validate()', () => {
      const t = new DateType({ min: 1, max: 10 })
      it('should validate a value and do not raise error', () => {
        expect(t.validate(2)).toBeTrue()
        expect(t.validate(0, false)).toBeFalse()
        expect(t.validate(11, false)).toBeFalse()
      })
      it('should validate a value and raise error', () => {
        expect(t.validate.bind(t, 0)).toThrow('is an invalid')
        expect(t.validate.bind(t, 11)).toThrow('is an invalid')
      })
      it('should validate a valid date', () => {
        expect(t.validate(new Date(5))).toBeTrue()
      })
      it('should validate a valid date string', () => {
        expect(t.validate(new Date(5).toISOString())).toBeTrue()
        // # because the Date to string ignores the micro-seconds!
        // # So the new Date(5).toString()->toDate() should be 0 now!!
        expect(t.validate(new Date(5).toString(), false)).toBeFalse()
      })

      it('should validate another DateType', () => {
        const today = DateType.today()
        const nextDay = today.clone().addDays(1)
        const lastDay = today.clone().addDays(-1)
        const FutureType = new Type(null, { name: 'date', min: nextDay })
        expect(FutureType.isValid(lastDay)).toBeFalse()
      })
    })
    describe('isLeapYear', () => {
      it('should check whether be a leap year', () => {
        expect(new DateType('2012').isLeapYear()).toBeTrue()
        expect(new DateType('2016').isLeapYear()).toBeTrue()
        expect(new DateType('2020').isLeapYear()).toBeTrue()
        expect(new DateType('2024').isLeapYear()).toBeTrue()

        for (const i of [2013, 2014, 2015]) {
          expect(new DateType(String(i)).isLeapYear()).toBeFalse()
        }

        for (const i of [2017, 2018, 2019]) {
          expect(new DateType(String(i)).isLeapYear()).toBeFalse()
        }

        for (const i of [2021, 2022, 2023]) {
          expect(new DateType(String(i)).isLeapYear()).toBeFalse()
        }
      })
    })

    describe('getDaysInMonth', () => {
      it('should get the days in a specified month', () => {
        expect(new DateType('2016-1').getDaysInMonth()).toEqual(31)
        expect(new DateType('2016-2').getDaysInMonth()).toEqual(29)
        expect(new DateType('2013-2').getDaysInMonth()).toEqual(28)
      })
    })

    describe('clearTime', () => {
      it('should get the days in a specified month', () => {
        const d = new DateType(new Date()).clearTime()
        expect(d.getHours()).toEqual(0)
        expect(d.getMinutes()).toEqual(0)
        expect(d.getSeconds()).toEqual(0)
        expect(d.getMilliseconds()).toEqual(0)
      })
    })

    describe('addDays', () => {
      it('should add some days to a date', () => {
        const d = new DateType('2016-12-1').clearTime().addDays(12)
        expect(d.equals(new DateType('2016-12-13').clearTime())).toBeTrue()
      })
    })
    describe('addWeeks', () => {
      it('should add some weeks to a date', () => {
        const d = new DateType('2016-2-28').clearTime().addWeeks(1)
        expect(d.equals(new DateType('2016-3-6').clearTime())).toBeTrue()
      })
    })
    describe('addMonths', () => {
      it('should add some months to a date', () => {
        const d = new DateType('2016-2-28').clearTime().addMonths(1)
        expect(d.equals(new DateType('2016-3-28').clearTime())).toBeTrue()
      })
    })
    describe('addYears', () => {
      it('should add some years to a date', () => {
        const d = new DateType('2016-2-28').clearTime().addYears(1)
        expect(d.equals(new DateType('2017-2-28').clearTime())).toBeTrue()
      })
    })
    describe('compareTo', () => {
      it('should compareTo', () => {
        expect(new DateType(2).compareTo(1)).toEqual(1)
        expect(new DateType(1).compareTo(2)).toEqual(-1)
        expect(new DateType(1).compareTo(1)).toEqual(0)
      })
    })

    describe('toObject', () => {
      it('should json a date type', () => {
        const dt = DateType.now()
        expect(JSON.stringify(dt)).toStrictEqual('"' + dt.toISOString() + '"')
      })
    })
    describe('toString', () => {
      it('should cast date type to string', () => {
        const dt = DateType.now()
        expect(String(dt)).toStrictEqual(dt.toISOString())
      })
    })
  })
})


.babelrc.js
const {
  engines: { node },
} = require('./package.json')
// require("@babel/register")({extensions: ['.js', '.ts']})
const presets = [
  // If you are not using typescript remove the following line.
  // '@babel/preset-typescript',
  // If you are not using flow remove the following line.
  // '@babel/preset-flow',
  [
    '@babel/env',
    {
      // Require the highest node version you can.
      // You should use at least a node version which
      // supports async/await because Babel has been
      // configured without polyfills/generators for
      // async/await.
      targets: {
        node: node.substring(2), // Strip `>=`
      },
    },
  ],
]

const plugins = [
  ['@babel/plugin-transform-typescript', { allowDeclareFields: true }],
  '@babel/plugin-proposal-class-properties',
  '@babel/plugin-proposal-object-rest-spread',
  ['@babel/plugin-proposal-decorators', { legacy: true }],
  [
    'module-resolver',
    {
      root: ['./src'],
      alias: {
        '@lib': './src/lib',
      },
    },
  ],
]

module.exports = {
  presets,
  plugins,

  // ignore: [
  //   '__tests__','__mocks__',
  //   '**/*.spec.js','**/*.test.js',
  //   '**/*.spec.ts','**/*.test.ts',
  // ],

  // For compatibility we generate inline source maps _and_
  // source maps in dedicated files. However due to a bug in babel
  // this option is not honored at the moment.
  sourceMaps: 'both',

  // Retaining lines increases debug-ability but may lead to less
  // readable source code. However if you have to debug for whatever
  // reason it is better to have the lines match up than to have
  // nicer source code.
  retainLines: true,
}


src/date-type.ts
/* eslint-disable no-unused-vars */
import isDate from 'util-ex/lib/is/type/date'

import {
  AbstractNumberType,
  // register,
  // defineProperties,
  ITypeObjectOptions,
} from 'number-type'

export class DateType extends AbstractNumberType {
  declare toDateString: () => string
  declare toTimeString: () => string
  declare toISOString: () => string
  declare toUTCString: () => string
  declare toGMTString: () => string
  declare getDate: Function
  declare setDate: Function
  declare getDay: Function
  declare getFullYear: Function
  declare setFullYear: Function
  declare getHours: Function
  declare setHours: Function
  declare getMilliseconds: Function
  declare setMilliseconds: Function
  declare getMinutes: Function
  declare setMinutes: Function
  declare getMonth: Function
  declare setMonth: Function
  declare getSeconds: Function
  declare setSeconds: Function
  declare getTime: Function
  declare setTime: Function
  declare getTimezoneOffset: Function
  declare getUTCDate: Function
  declare setUTCDate: Function
  declare getUTCDay: Function
  declare getUTCFullYear: Function
  declare setUTCFullYear: Function
  declare getUTCHours: Function
  declare setUTCHours: Function
  declare getUTCMilliseconds: Function
  declare setUTCMilliseconds: Function
  declare getUTCMinutes: Function
  declare setUTCMinutes: Function
  declare getUTCMonth: Function
  declare setUTCMonth: Function
  declare getUTCSeconds: Function
  declare setUTCSeconds: Function
  declare getYear: Function
  declare setYear: Function
  declare toLocaleString: () => string
  declare toLocaleDateString: () => string
  declare toLocaleTimeString: () => string
  declare clone: (options?: ITypeObjectOptions) => DateType

  declare static min: number | undefined
  declare static max: number | undefined
  declare min: number | undefined
  declare max: number | undefined
  // declare _min: number | undefined
  // declare _max: number | undefined

  static toValue(aValue): number | undefined {
    let result: number | undefined =
      aValue != null && typeof aValue !== 'number'
        ? new Date(aValue).valueOf()
        : aValue

    if (isNaN(result)) result = undefined

    return result
  }

  /**
   * Determines if the current date instance is within a LeapYear.
   * @param aYear The year.
   * @returns true if date is within a LeapYear, otherwise false.
   */
  static isLeapYear(aYear: number): boolean {
    return (aYear & 3) === 0 && (aYear % 25 !== 0 || (aYear & 15) === 0)
  }

  /**
   * Gets the number of days in the month, given a year and month value. Automatically corrects for LeapYear.
   * @param aYear  The year.
   * @param aMonth The month (0-11).
   * @returns The number of days in the month.
   */
  static getDaysInMonth(aYear: number, aMonth: number): number {
    return [
      31,
      this.isLeapYear(aYear) ? 29 : 28,
      31,
      30,
      31,
      30,
      31,
      31,
      30,
      31,
      30,
      31,
    ][aMonth]
  }

  /**
   * Compares the first date to the second date and returns an number indication of their relative values.
   * @param value1 First Date object to compare [Required].
   * @param value2 Second Date object to compare to [Required].
   * @returns -1 = value1 is less than value2. 0 = values are equal. 1 = value1 is greater than value2.
   */
  static compare(value1: Date | string, value2: Date | string): -1 | 0 | 1 {
    const v1 = new Date(this.toValue(value1) as number)
    const v2 = new Date(this.toValue(value2) as number)
    if (!isNaN(v1.valueOf()) && !isNaN(v2.valueOf())) {
      return v1 < v2 ? -1 : v1 > v2 ? 1 : 0
    } else {
      throw new TypeError(value1 + ' - ' + value2)
    }
  }

  /**
   * Compares the first Date object to the second Date object and returns true if they are equal.
   * @param value1 First Date object to compare [Required].
   * @param value2 Second Date object to compare to [Required].
   * @returns true if dates are equal. false if they are not equal.
   */
  static equals(value1, value2): boolean {
    return this.compare(value1, value2) === 0
  }

  /**
   * Resets the time of the Date object to 12:00 AM (00:00), which is the start of the day.
   * @param aValue clearing time for this date instance
   * @returns the aValue date instance
   */
  static clearTime(aValue: Date | DateType) {
    aValue.setHours(0)
    aValue.setMinutes(0)
    aValue.setSeconds(0)
    aValue.setMilliseconds(0)
    return aValue
  }

  static today(aOptions?: ITypeObjectOptions) {
    const result = this.now(aOptions)
    this.clearTime(result)
    return result
  }

  static toISOString(aValue) {
    /* istanbul ignore else */
    if (!isDate(aValue) || !(aValue instanceof DateType)) {
      aValue = new Date(aValue)
    }
    return aValue.toISOString()
  }

  static now(aOptions?: ITypeObjectOptions): DateType {
    return new DateType(new Date(), aOptions)
  }

  static parse = Date.parse
  static UTC = Date.UTC

  // _initialize(aValue?, aOptions: ITypeObjectOptions) {
  //   if (
  //     (aValue == null || aValue.value == null) &&
  //     aOptions.value === undefined
  //   ) {
  //     aOptions.value = new Date()
  //   }
  // }

  _toObject() {
    return this.toISOString()
  }

  toString() {
    return this.toISOString()
  }

  isLeapYear() {
    return DateType.isLeapYear(this.getFullYear())
  }

  getDaysInMonth() {
    return DateType.getDaysInMonth(this.getFullYear(), this.getMonth())
  }

  /**
   * Resets the time of this Date object to 12:00 AM (00:00), which is the start of the day.
   * @returns this date instance
   */
  clearTime() {
    return DateType.clearTime(this)
  }

  /**
   * Adds the specified number of days to this instance.
   * @param value The number of days to add. The number can be positive or negative [Required]
   * @returns this
   */
  addDays(value: number) {
    this.setDate(this.getDate() + value * 1)
    return this
  }

  /**
   * Adds the specified number of weeks to this instance.
   * @param value The number of weeks to add. The number can be positive or negative [Required]
   * @returns this
   */
  addWeeks(value: number) {
    return this.addDays(value * 7)
  }

  /**
   * Adds the specified number of months to this instance.
   * @param value The number of months to add. The number can be positive or negative [Required]
   * @returns this
   */
  addMonths(value) {
    const n = this.getDate()
    this.setDate(1)
    this.setMonth(this.getMonth() + value)
    this.setDate(Math.min(n, this.getDaysInMonth()))
    return this
  }

  /**
   * Adds the specified number of years to this instance.
   * @param value The number of years to add. The number can be positive or negative [Required]
   * @returns this
   */
  addYears(value) {
    return this.addMonths(value * 12)
  }

  /**
   * Compares this instance to another Date object and returns true if they are equal.
   * @param value Date object to compare. If no date to compare, new Date() [now] is used.
   * @returns true if dates are equal. false if they are not equal.
   */
  equals(value) {
    return DateType.equals(this.value, value)
  }

  /**
   * Compares this instance to a Date object and returns an number indication of their relative values.
   * @param value Date object to compare [Required]
   * @returns -1 = this is less than date. 0 = values are equal. 1 = this is greater than date.
   */
  compareTo(value) {
    return DateType.compare(this.value, value)
  }
}

const vDateMethodNames = Object.getOwnPropertyNames(Date.prototype).filter(
  (name) =>
    ['constructor', 'valueOf', 'toJSON', 'toString'].indexOf(name) === -1
)

vDateMethodNames.forEach((name) => {
  DateType.prototype[name] = (function (aName) {
    return function () {
      /* istanbul ignore else */
      if (this.value != null) {
        const v = new Date(this.value)
        const result = Date.prototype[aName].apply(v, arguments)
        if (aName.slice(0, 3) === 'set') {
          this.value = v.valueOf()
        }
        return result
      }
    }
  })(name)
})

AbstractNumberType.register(DateType, {
  aliases: ['date', 'time', 'Time', 'DateTime', 'datetime'],
})
